src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1029 @@
+	{
+		place_cursor(m_cursor_row, m_cursor_col+1);
+	}
+
+	m_speed_check = true;
+	m_speed_ctrl->adjust(attotime::from_usec(16121));
+}
+
+void tandberg_tdv2100_disp_logic_device::place_cursor(int row, int col)
+{
+	bool in_eol_region = ((m_cursor_col&0x48) == 72);
+
+	m_cursor_col = col&0x7f;
+	m_cursor_row = row&0x1f;
+
+	if((m_cursor_col&0x48) == 72 && !in_eol_region && !m_speed_check)
+	{
+		// End of line warning beep
+		char_to_display(0x07);
+	}
+	if((m_cursor_col&0x50) == 80)
+	{
+		// End of line action
+		char_to_display(0x0d);
+		char_to_display(0x0a);
+	}
+}
+
+int tandberg_tdv2100_disp_logic_device::get_ram_addr(int row, int col)
+{
+	int abs_row = (row&0x1f) + m_page_roll;
+	return ((abs_row<<7) + (col&0x7f) + (m_addr_offsets[abs_row]<<4))&0x7ff;
+}
+
+void tandberg_tdv2100_disp_logic_device::erase_row(int row)
+{
+	for(int col = 0; col<80; col++)
+	{
+		m_vram[get_ram_addr(row, col)] = 0;
+	}
+}
+
+TIMER_CALLBACK_MEMBER(tandberg_tdv2100_disp_logic_device::end_beep)
+{
+	m_beep->set_state(0);
+}
+
+TIMER_CALLBACK_MEMBER(tandberg_tdv2100_disp_logic_device::expire_speed_check)
+{
+	m_speed_check = false;
+}
+
+void tandberg_tdv2100_disp_logic_device::vblank(int state)
+{
+	if(state && !m_vblank_state)
+	{
+		// TODO: VSync interrupt for CPU interface
+		m_frame_counter++;
+		m_screen->update_now();
+	}
+	m_vblank_state = state;
+}
+
+int tandberg_tdv2100_disp_logic_device::get_attribute(int at_row, int at_col)
+{
+	int attribute = 0;
+	for(int char_row = 0; char_row <= at_row; char_row++)
+	{
+		for(int char_col = 0; char_col < 80; char_col++)
+		{
+			int const vram_data = m_vram[char_col + char_row*80];
+			if(vram_data&0x80)
+			{
+				attribute = (vram_data&0x70)>>4;
+			}
+			if(char_row == at_row && char_col == at_col)
+			{
+				break;
+			}
+		}
+	}
+	return attribute;
+}
+
+uint32_t tandberg_tdv2100_disp_logic_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	rgb_t const *const palette = m_palette->palette()->entry_list_raw();
+
+	bool extend_dot7 = ((m_dsw_u61->read()&0x003) == 1);
+	bool blinking_cursor = ((m_dsw_u61->read()&0x030) == 0x010);
+	bool blank_attr_chars = !(m_dsw_u61->read()&0x040);
+	bool blank_ctrl_chars = !(m_dsw_u61->read()&0x100);
+	bool block_cursor = m_dsw_u61->read()&0x200;
+	bool full_inverse_video = !m_sw_invert_video->read();
+
+	int const char_min_col = cliprect.min_x/9;
+	int const char_max_col = cliprect.max_x/9;
+	int const char_min_row = cliprect.min_y/14;
+	int const char_max_row = cliprect.max_y/14;
+	bool const blink_strobe = m_frame_counter&0x08;
+	bool const cursor_blink_strobe = blink_strobe && !m_speed_check;
+
+	for(int char_row_nr = char_min_row; char_row_nr <= char_max_row; char_row_nr++)
+	{
+		// This is at the start of a char row
+		int const chr_y_pos = char_row_nr*14;
+		int attribute = get_attribute(char_row_nr, char_min_col);
+
+		for (int char_col_nr = char_min_col; char_col_nr <= char_max_col; char_col_nr++)
+		{
```

Comment:
This `for` statement has different spacing to the others.

Reply:
Fixed.